File: Daily Resources/Day13/DSA - Two Pointers Solutions/python/Trapping Rain Water.py
from typing import List
def trap(height: List[int]) -> int:
    left, right = 0, len(height) - 1
    left_max = right_max = res = 0
    while left < right:
        if height[left] < height[right]:
            left_max = max(left_max, height[left])
            res += left_max - height[left]
            left += 1
        else:
            right_max = max(right_max, height[right])
            res += right_max - height[right]
            right -= 1
    return res

File: Daily Resources/Day13/DSA - Two Pointers Solutions/cpp/Trapping Rain Water.cpp
#include <bits/stdc++.h>
using namespace std;
int trap(vector<int>& height) {
    int left = 0, right = height.size() - 1, leftMax = 0, rightMax = 0, res = 0;
    while (left < right) {
        if (height[left] < height[right]) {
            leftMax = max(leftMax, height[left]);
            res += leftMax - height[left++];
        } else {
            rightMax = max(rightMax, height[right]);
            res += rightMax - height[right--];
        }
    }
    return res;
}
